src/error_utilities.rs

```diff
@@ -0,0 +1,44 @@
+use std::io;
+
+use tracing::error;
+
+/// Get the last os error, log with note and return the error
+pub(crate) fn log_ret_last_os_err_with_note(func_name: &str, note: &str) -> io::Error {
+    let err = io::Error::last_os_error();
+    if note.is_empty() {
+        error!("{} ret os error {:?}", func_name, err);
+    } else {
+        error!("{} ret os error {:?}. Note: {}", func_name, err, note);
+    }
+    err
+}
+
+/// Get the last os error, log and return the error
+pub(crate) fn log_ret_last_os_err(func_name: &str) -> io::Error {
+    log_ret_last_os_err_with_note(func_name, "")
+}
+
+/// Get the last os error and just log it
+pub(crate) fn log_last_os_err(func_name: &str) {
+    #[allow(clippy::pedantic)] // some errors occur in `drop` methord is not necessary to panic
+    let _ = log_ret_last_os_err_with_note(func_name, "");
+}
+
+/// Get the name of a function
+#[macro_export]
+macro_rules! func_name {
+    () => {{
+        fn f() {}
+        fn type_name_of<T>(_: T) -> &'static str {
+            std::any::type_name::<T>()
+        }
+        let name = type_name_of(f);
+        &name[..name.len().saturating_sub(3)]
+    }};
+}
```

Comment:
I have seen this before. It's better to put a link for explanation. 
<https://stackoverflow.com/a/40234666>

Reply:
Yes, I copy that macro from there and forget ref it, thanks for reminding.
Do you think this macro is good? I didn't find official support or stable library.

Reply:
An interesting hack. 
I personally prefer not to use the hack because the log and [stack backtrace](https://docs.rs/backtrace/latest/backtrace/) already have enough information. Sometimes the [span trace](https://docs.rs/tracing-error/latest/tracing_error/struct.SpanTrace.html) can provide more precise debug info.